File: client/lib/net/host_session.dart
/// Host side of a match — twin of `net/HostSession.ts`. Solo play is a
/// HostSession with no transport ("host with zero peers"). Accepts joiners,
/// hands them the world diff, applies their block edits and broadcasts
/// snapshots at [snapshotHz]. Every joiner gets an [Avatar] in the game's
/// [HostSim], so the host simulates their vitals, inventory and combat and
/// sends each one its private state.
library;

import 'dart:async';
import 'dart:typed_data';

import 'package:block_survival/api/api_client.dart';
import 'package:block_survival/api/game_api.dart';
import 'package:block_survival/game/avatar.dart';
import 'package:block_survival/game/game.dart';
import 'package:block_survival/game/score.dart';
import 'package:block_survival/game/ui_state.dart';
import 'package:block_survival/net/protocol.dart';
import 'package:block_survival/net/rtc.dart';
import 'package:block_survival/net/transport.dart';
import 'package:block_survival/world/seed.dart';

/// how often the host refreshes its room row on the API (seconds); in the
/// global world this is also the heartbeat that keeps everyone's seat
const double roomRefreshSeconds = 15;

final class HostSession implements TransportEvents {
  HostSession(this._api, this._rtc, {String? code})
    : code = code ?? makeRoomCode();

  final GameApi _api;
  final RtcFactory _rtc;
  final String code;
  HostTransport? _transport;
  Game? _game;
  WorldKind _worldKind = WorldKind.own;
  final Set<String> _pending = {};
  final Map<String, Avatar> players = {};
  final Map<String, Link> _links = {};
  final Map<String, int> _inventorySent = {};
  final List<BlockEdit> _outgoing = [];
  double _snapshotTimer = 0;
  double _roomRefreshTimer = 0;

  /// the server no longer counts us as the global world's host (its queue moved on)
  void Function(String reason)? onLost;
  void Function()? onPlayersChanged;

  bool get online => _transport != null;
  int get playerCount => players.length + 1;

  void attach(Game game) {
    _game = game;
    game.onEdit = _outgoing.add;
    game.remotePlayers = const [];
  }

  /// Open the room: register the code → host id with the app so joiners can
  /// resolve it, then start reading the room's mailbox.
  Future<void> listen(String hostName, WorldKind worldKind) async {
    if (_transport != null) return;
    final t = HostTransport(_api, _rtc, this);
    await _api.createRoom(code, t.id, hostName, worldKind: worldKind);
    t.listen(code);
    _transport = t;
    _worldKind = worldKind;
  }

  void tick(double dt) {
    final game = _game;
    final t = _transport;
    if (game == null || t == null) return;
    _roomRefreshTimer += dt;
    if (_roomRefreshTimer >= roomRefreshSeconds) {
      _roomRefreshTimer = 0;
      final userIds = [
        if (game.local.userId != null) game.local.userId!,
        for (final p in players.values)
          if (p.userId != null) p.userId!,
      ];
      unawaited(
        _api
            .refreshRoom(code, t.id, playerCount, userIds: userIds)
            .catchError(_onRefreshFailed),
      );
    }
    if (_outgoing.isNotEmpty) {
      t.broadcast(encodeHost(Blocks(List.of(_outgoing))));
      _outgoing.clear();
    }
    _snapshotTimer += dt;
    if (_snapshotTimer >= 1 / snapshotHz) {
      _snapshotTimer = 0;
      t.broadcast(encodeHost(game.sim.snapshot()));
      _sendPrivateState();
      _refreshRows(game);
    }
  }

  /// each client gets what only it may know: its inventory when it changed,
  /// plus whatever the sim queued (vitals, magazine, teleports, messages)
  void _sendPrivateState() {
    for (final p in players.values) {
      final v = p.inventory.version;
      if (_inventorySent[p.id] != v) {
        _inventorySent[p.id] = v;
        p.push(inventory: p.inventory.all());
      }
      final state = p.takeOutbox();
      if (state != null) _links[p.id]?.send(encodeHost(state));
    }
  }

  void _onRefreshFailed(Object e) {
    // a refresh the server refused outright means the global world has a new host
    if (_worldKind != WorldKind.global || e is! ApiError) return;
    if (e.status == 404 || e.status == 409) onLost?.call(e.message);
  }

  @override
  void onOpen(Link link) => _pending.add(link.id);

  @override
  void onData(Link link, Uint8List bytes) {
    final game = _game;
    if (game == null) return;
    final ClientMessage msg;
    try {
      msg = decodeClient(bytes);
    } on Object {
      return;
    }
    if (_pending.contains(link.id)) {
      if (msg is! Hello || msg.v != protocolVersion) return;
      _pending.remove(link.id);
      if (playerCount >= maxPlayers) {
        link.send(encodeHost(const Full()));
        unawaited(link.close());
        return;
      }
      final name = msg.name.isEmpty
          ? 'Player'
          : msg.name.substring(0, msg.name.length > 16 ? 16 : msg.name.length);
      final player = Avatar(
        id: link.id,
        name: name,
        userId: msg.userId,
        spawn: game.spawn,
      );
      final saved = game.savedVisitor(msg.userId);
      if (saved != null) Game.restorePlayer(player, saved, pose: false);
      players[link.id] = player;
      _links[link.id] = link;
      game.sim.avatars[link.id] = player;
      link.send(
        encodeHost(
          Welcome(
            v: protocolVersion,
            you: link.id,
            seed: game.seed,
            time: game.time,
            edits: game.world.edits.values
                .map((e) => BlockEdit(x: e.x, y: e.y, z: e.z, id: e.id))
                .toList(),
            spawn: game.spawn,
            rules: game.rules,
          ),
        ),
      );
      _inventorySent[link.id] = player.inventory.version;
      link.send(
        encodeHost(
          PrivateState(
            inventory: player.inventory.all(),
            magazine: player.magazine,
            health: player.health,
            spawn: player.spawn,
          ),
        ),
      );
      game.toast('$name joined');
      _playersChanged(game);
      return;
    }
    final player = players[link.id];
    if (player == null) return;
    switch (msg) {
      case InputMessage(
        :final x,
        :final y,
        :final z,
        :final yaw,
        :final pitch,
        :final anim,
        :final slot,
        :final aiming,
      ):
        player
          ..x = x
          ..y = y
          ..z = z
          ..yaw = yaw
          ..pitch = pitch
          ..anim = anim
          ..slot = slot
          ..aiming = aiming;
      case ClientChat(:final text):
        _transport?.broadcast(encodeHost(HostChat(player.name, text)));
        game.toast('${player.name}: $text');
      case Hello():
        break;
      default:
        game.sim.apply(player, msg);
    }
  }

  @override
  void onClose(Link link) {
    _pending.remove(link.id);
    final player = players.remove(link.id);
    _links.remove(link.id);
    _inventorySent.remove(link.id);
    final game = _game;
    if (player != null && game != null) {
      game.sim.avatars.remove(link.id);
      game.leaveSaved(player);
      game.toast('${player.name} left');
      _playersChanged(game);
    }
  }

  void _playersChanged(Game game) {
    _refreshRows(game);
    onPlayersChanged?.call();
  }

  void _refreshRows(Game game) {
    game.remotePlayers = [
      for (final p in players.values)
        ScoreRow(
          id: p.id,
          name: p.name,
          score: computeScore(game.nightsSurvived, p.kills),
          kills: p.kills,
          deaths: p.deaths,
          you: false,
        ),
    ];
  }

  Future<void> dispose() async {
    final t = _transport;
    if (t != null) {
      t.broadcast(encodeHost(const Bye()));
      unawaited(_api.closeRoom(code, t.id).catchError((Object _) {}));
      await t.dispose();
    }
    _transport = null;
    _game?.onEdit = null;
  }
}

File: server/resources/js/net/HostSession.ts
/**
 * Host side of a match. Solo play is a HostSession with no transport ("host with zero
 * peers"). Accepts joiners, hands them the world diff, feeds their messages to the
 * authoritative Game and broadcasts snapshots at SNAPSHOT_HZ.
 */
import type { Game } from '../game/Game'
import { rulesToWire } from '../game/rules'
import type { WorldKind } from '../world/seed'
import { HostTransport, type Link } from './transport'
import {
  decode, encode, makeRoomCode, MAX_PLAYERS, PROTOCOL_VERSION, SNAPSHOT_HZ,
  type ClientMessage, type HostMessage, type Welcome,
} from './protocol'
import { api, ApiError } from './api'

/**
 * how often the host refreshes its room row on the API (seconds) — in the global world
 * this is also the heartbeat that keeps everyone's seat (a seat is stale after 45 s)
 */
export const ROOM_REFRESH_SECONDS = 15

export class HostSession {
  readonly role = 'host' as const
  readonly code: string
  private transport: HostTransport | null = null
  private game: Game | null = null
  private snapshotTimer = 0
  /** links that connected but have not said hello yet */
  private readonly pending = new Set<string>()
  onPlayersChanged: (() => void) | null = null
  /** the server no longer counts us as the global world's host (its queue moved on) */
  onLost: ((reason: string) => void) | null = null
  private roomRefreshTimer = 0
  private worldKind: WorldKind = 'own'

  constructor(code = makeRoomCode()) {
    this.code = code
  }

  attach(game: Game): void {
    this.game = game
  }

  get playerCount(): number {
    return this.game ? this.game.avatars.size : 1
  }

  get online(): boolean {
    return this.transport !== null
  }

  /**
   * Open the room: register the code → host id with the app so joiners can resolve
   * it, then start reading the room's mailbox (which 404s until the room exists).
   * Fails if the app cannot be reached.
   */
  async listen(hostName = 'Survivor', worldKind: WorldKind = 'own'): Promise<void> {
    if (this.transport) return
    const t = new HostTransport({
      onOpen: link => { this.pending.add(link.id) },
      onData: (link, bytes) => this.onData(link, bytes),
      onClose: link => this.onLeave(link),
    })
    await api.createRoom(this.code, t.id, hostName, worldKind)
    await t.listen(this.code)
    this.transport = t
    this.worldKind = worldKind
  }

  tick(dt: number): void {
    const game = this.game
    if (!game || !this.transport) return
    this.roomRefreshTimer += dt
    if (this.roomRefreshTimer >= ROOM_REFRESH_SECONDS) {
      this.roomRefreshTimer = 0
      const userIds = [...game.avatars.values()].map(a => a.userId).filter((id): id is number => id !== null)
      api.refreshRoom(this.code, this.transport.id, game.avatars.size, userIds).catch(e => this.onRefreshFailed(e))
    }
    // block edits go out immediately, snapshots at a fixed rate, private state when dirty
    const edits = game.takeBlockEdits()
    if (edits.length) this.transport.broadcast(encode({ t: 'blocks', edits }))
    this.snapshotTimer += dt
    if (this.snapshotTimer >= 1 / SNAPSHOT_HZ) {
      this.snapshotTimer = 0
      this.transport.broadcast(encode(game.buildSnapshot()))
    }
    for (const avatar of game.avatars.values()) {
      if (avatar === game.local) continue
      const link = this.transport.links.get(avatar.id)
      if (!link) continue
      if (avatar.inventory.version !== avatar.inventoryVersionSent) {
        avatar.inventoryVersionSent = avatar.inventory.version
        avatar.push({ inventory: avatar.inventory.all() })
      }
      const state = avatar.takeOutbox()
      if (state) link.send(encode(state))
    }
  }

  /** a refresh the server refused outright means the global world has a new host; anything else is retried next time */
  private onRefreshFailed(e: unknown): void {
    if (this.worldKind !== 'global' || !(e instanceof ApiError) || (e.status !== 404 && e.status !== 409)) return
    this.onLost?.(e.message)
  }

  dispose(): void {
    if (this.transport) {
      this.transport.broadcast(encode({ t: 'bye' }))
      api.closeRoom(this.code, this.transport.id).catch(() => {})
    }
    this.transport?.dispose()
    this.transport = null
  }

  private send(link: Link, msg: HostMessage): void {
    link.send(encode(msg))
  }

  private onData(link: Link, bytes: Uint8Array): void {
    const game = this.game
    if (!game) return
    let msg: ClientMessage
    try {
      msg = decode<ClientMessage>(bytes)
    } catch {
      return
    }
    if (this.pending.has(link.id)) {
      if (msg.t !== 'hello' || msg.v !== PROTOCOL_VERSION) return
      this.pending.delete(link.id)
      if (game.avatars.size >= MAX_PLAYERS) {
        this.send(link, { t: 'full' })
        link.close()
        return
      }
      const avatar = game.addRemoteAvatar(link.id, msg.name.slice(0, 16) || 'Player', typeof msg.userId === 'number' ? msg.userId : null)
      const welcome: Welcome = {
        t: 'welcome', v: PROTOCOL_VERSION, you: avatar.id, seed: game.seed, time: game.dayNight.time,
        edits: game.worldEdits(), spawn: avatar.spawn, rules: rulesToWire(game.rules),
      }
      this.send(link, welcome)
      avatar.push({ inventory: avatar.inventory.all(), magazine: avatar.magazine, health: avatar.health, spawn: avatar.spawn })
      game.showMessage(`${avatar.name} joined`)
      this.onPlayersChanged?.()
      return
    }
    const avatar = game.avatars.get(link.id)
    if (!avatar) return
    game.applyClientMessage(avatar, msg)
  }

  private onLeave(link: Link): void {
    this.pending.delete(link.id)
    const game = this.game
    if (!game) return
    const avatar = game.avatars.get(link.id)
    if (avatar) {
      game.removeAvatar(link.id)
      game.showMessage(`${avatar.name} left`)
      this.onPlayersChanged?.()
    }
  }
}
